### PR TITLE
Add a sfizz_render client in-tree

### PR DESCRIPTION
**Pull request by paulfd**
This uses the `fmidi` library and its player, which adds a dependency on `libev`. The goal of having this client in-tree is to facilitate benchmarking of different versions.

@jpcima not sure I use your library right so check it :slightly_smiling_face:

### CODE REVIEW

**Comment by jpcima** [comment]
>This uses the fmidi library and its player, which adds a dependency on libev. 

This is not true, it does not add a dependency on libev.
Only the example programs are dependent of the library, libev was moved out of the main library since a long time.

The library depends on `boost` and on `fmt`, and that's only for the description API which converts SMF to plain text.

I have around an edit of the library that can disable the above feature, such that it does not have external deps. Then it's a ~140K source code without recursive submodules.
https://github.com/jpcima/midiplay/tree/master/thirdparty/fmidi
https://github.com/jpcima/midiplay/blob/21c095fdf6b8326dc00608046e9ccf6f28b734a2/Makefile#L77

**Comment by paulfd** [comment]
It seems I had to install libev to access the `fmidi_player_XX` interfaces? After some more reading it seems to be unneeded. There are these lines in the CMakeLists
```cmake
find_library(ev_LIBRARY ev)
set(fmidi-player_BUILD FALSE)
if(ev_LIBRARY)
  set(fmidi-player_BUILD TRUE)
else()
  message(STATUS "libev is missing, NOT building library fmidi-player.")
endif()
```

File: clients/sfizz_render.cpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+void midiCallback(const fmidi_event_t * event, void * cbdata)
+{
+    auto data = reinterpret_cast<CallbackData*>(cbdata);
+
+    if (event->type != fmidi_event_type::fmidi_event_message)
+        return;
+
+    switch (midi::status(event->data[0])) {
+        case midi::noteOff:
+            data->synth.noteOff(event->delta, event->data[1], event->data[2]);
```
Comment (jpcima) [inline clients/sfizz_render.cpp]: The field `event->delta` is not relevant to sequencing based on delta times, it's a raw MIDI delta value (related to tempo).

For now, the way to go is to force delta as 0, and process in small buffer sizes. (in the order of 5ms)

Ideally the library should pass the player a timestamp, which right now is not the case.

**Comment by jpcima** [comment]
>It seems I had to install libev to access the fmidi_player_XX interfaces?

It's the example player that comes with as part of the library. (see `programs/`)

File: clients/sfizz_render.cpp (hunk 1 above)
Comment (jpcima) [reply clients/sfizz_render.cpp]: In fact, a much better idea, after thinking better about it:

Keep the block size large, like it is.
Tick the midi player by 1/sampleRate for each frame, and increment a counter each time in the callback context, that starts at 0 for each block.
Use this counter as MIDI delay to sfizz events.

**Comment by paulfd** [comment]
If we go with #201 can I close this one safely?

**Comment by jpcima** [comment]
Yes, it's no problem to close this one.
It's based on the identical commit ref 6f39d4c.

**Pull request closed without merging**